File: modules/accessibility/checks/sr_names.py
"""Normalise + detect missing accessible-name issues.

The accessible name for an interactive element is computed from the
ARIA chain (``aria-labelledby`` → ``aria-label`` → label text →
visible content → ``title`` → placeholder). When every fallback is
empty the element has no accessible name and screen readers cannot
announce it.

the documentation calls this out explicitly; the curated handful of rules
below is the deterministic Python mirror of the TS helper.
"""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass
from typing import Any

from modules.accessibility.models import AccessibleNameIssue

INTERACTIVE_ROLES: frozenset[str] = frozenset(
    {"button", "link", "textbox", "combobox", "checkbox", "radio", "switch", "menuitem"}
)


@dataclass(frozen=True)
class ElementSnapshot:
    """Minimal subset of a DOM element needed to compute an accessible name."""

    role: str
    selector: str
    aria_label: str = ""
    aria_labelledby_text: str = ""
    label_text: str = ""
    visible_text: str = ""
    title: str = ""
    placeholder: str = ""


def normalise_accessible_name_issues(
    raw: Iterable[Any],
) -> tuple[AccessibleNameIssue, ...]:
    """Coerce a sequence of raw dicts into typed records."""

    out: list[AccessibleNameIssue] = []
    for entry in raw:
        if not isinstance(entry, dict):
            continue
        selector = str(entry.get("selector") or "").strip()
        description = str(entry.get("description") or "").strip()
        if not selector or not description:
            continue
        out.append(
            AccessibleNameIssue(
                selector=selector,
                role=str(entry.get("role") or ""),
                description=description,
            )
        )
    return tuple(out)


def has_accessible_name(element: ElementSnapshot) -> bool:
    """Return True when the element carries a non-empty accessible name."""

    for value in (
        element.aria_labelledby_text,
        element.aria_label,
        element.label_text,
        element.visible_text,
        element.title,
    ):
        if value and value.strip():
            return True
    # `placeholder` is intentionally not a sufficient fallback —
    # CLAUDE §28 / the documentation: placeholders disappear on input, so
    # they fail accessible-name requirements.
    return False


def detect_missing_accessible_names(
    elements: Iterable[ElementSnapshot],
) -> tuple[AccessibleNameIssue, ...]:
    """Yield one issue per interactive element with no accessible name."""

    issues: list[AccessibleNameIssue] = []
    for element in elements:
        if element.role not in INTERACTIVE_ROLES:
            continue
        if has_accessible_name(element):
            continue
        issues.append(
            AccessibleNameIssue(
                selector=element.selector,
                role=element.role,
                description=(
                    f"{element.role!r} element has no computable accessible "
                    "name; screen readers cannot announce it."
                ),
            )
        )
    return tuple(issues)


__all__ = [
    "ElementSnapshot",
    "INTERACTIVE_ROLES",
    "detect_missing_accessible_names",
    "has_accessible_name",
    "normalise_accessible_name_issues",
]

File: packages/ts-runtime/src/a11y/sr_names.ts
// accessible-name check.
// Mirrors `modules/accessibility/checks/sr_names.py`: every interactive
// element must produce a computable accessible name. Placeholders are
// NOT a sufficient fallback because they
// disappear on input.

import type { AccessibleNameIssue } from './types.js';

export const INTERACTIVE_ROLES: ReadonlySet<string> = new Set([
  'button',
  'link',
  'textbox',
  'combobox',
  'checkbox',
  'radio',
  'switch',
  'menuitem',
]);

export interface ElementSample {
  readonly role: string;
  readonly selector: string;
  readonly ariaLabel: string;
  readonly ariaLabelledbyText: string;
  readonly labelText: string;
  readonly visibleText: string;
  readonly title: string;
  readonly placeholder: string;
}

export function hasAccessibleName(el: ElementSample): boolean {
  return Boolean(
    el.ariaLabelledbyText?.trim() ||
      el.ariaLabel?.trim() ||
      el.labelText?.trim() ||
      el.visibleText?.trim() ||
      el.title?.trim(),
  );
}

export function detectMissingAccessibleNames(
  elements: readonly ElementSample[],
): AccessibleNameIssue[] {
  const issues: AccessibleNameIssue[] = [];
  for (const el of elements) {
    if (!INTERACTIVE_ROLES.has(el.role)) continue;
    if (hasAccessibleName(el)) continue;
    issues.push({
      selector: el.selector,
      role: el.role,
      description: `Automated accessibility check found ${el.role} element with no computable accessible name; screen readers cannot announce it.`,
    });
  }
  return issues;
}

export interface SrNamesPageLike {
  evaluate<T>(fn: () => T | Promise<T>): Promise<T>;
}

export async function readInteractiveSamples(
  page: SrNamesPageLike,
): Promise<readonly ElementSample[]> {
  return await page.evaluate(() => {
    const roleOf = (el: Element): string => {
      const explicit = el.getAttribute('role');
      if (explicit) return explicit;
      const tag = el.tagName.toLowerCase();
      if (tag === 'button') return 'button';
      if (tag === 'a' && el.hasAttribute('href')) return 'link';
      if (tag === 'input') {
        const type = (el.getAttribute('type') ?? 'text').toLowerCase();
        if (type === 'checkbox') return 'checkbox';
        if (type === 'radio') return 'radio';
        if (type === 'button' || type === 'submit') return 'button';
        return 'textbox';
      }
      if (tag === 'select') return 'combobox';
      if (tag === 'textarea') return 'textbox';
      return '';
    };
    const labelText = (el: Element): string => {
      if (!(el instanceof HTMLElement)) return '';
      const id = el.id;
      if (id) {
        const lbl = document.querySelector(`label[for="${id}"]`);
        if (lbl !== null) return lbl.textContent?.trim() ?? '';
      }
      const parent = el.closest('label');
      return parent ? (parent.textContent?.trim() ?? '') : '';
    };
    const out: ElementSample[] = [];
    const candidates = document.querySelectorAll(
      'button, a[href], input, select, textarea, [role]',
    );
    candidates.forEach((el) => {
      const role = roleOf(el);
      if (!role) return;
      const selector =
        (el as HTMLElement).id !== '' ? `#${(el as HTMLElement).id}` : el.tagName.toLowerCase();
      const ariaLabelledby = el.getAttribute('aria-labelledby');
      let ariaLabelledbyText = '';
      if (ariaLabelledby) {
        const parts = ariaLabelledby
          .split(/\s+/)
          .map((id) => document.getElementById(id)?.textContent?.trim() ?? '')
          .filter(Boolean);
        ariaLabelledbyText = parts.join(' ');
      }
      out.push({
        role,
        selector,
        ariaLabel: el.getAttribute('aria-label') ?? '',
        ariaLabelledbyText,
        labelText: labelText(el),
        visibleText: (el.textContent ?? '').trim(),
        title: el.getAttribute('title') ?? '',
        placeholder: el.getAttribute('placeholder') ?? '',
      });
    });
    return out;
  });
}
